File: clojure/data_structures/union_find/quick_union.clj
(ns union-find.quick-union)

(defprotocol IQuickUnion
  (root [this p])
  (is-connected? [this p q])
  (union [this p q])
  (print-arr [this]))

(deftype UnionFind [xs]
  IQuickUnion
  (root [this p]
    (let [parent (get xs p)]
      (if (= parent p)
        p
        (recur parent))))
  (is-connected? [this p q]
    (= (root this p) (root this q)))

  (union [this p q]
    "Time complexity: O(n) - worst case
    Change root of p to point to root of q."
    (let [p-root (root this p)
          q-root (root this q)]
      (UnionFind.
       (assoc xs p-root q-root))))

  (print-arr [this]
    (mapv (fn [e] e) xs))

  #_java.lang.Object
  #_(toString [this])

  ;
  )

(defn quick-find [n]
  (->UnionFind (vec (range n))))

(comment
  (-> (quick-find 10)
      (union 1 2)
      (is-connected? 1 2))
  ;=> true

  (-> (quick-find 10)
      (union 4 3)
      (union 3 8)
      (union 6 5)
      (union 9 4)
      (union 2 1)
      (union 5 0)
      (union 7 2)
      (union 6 1)
      (union 7 3)
      (print-arr))

  ;=> [1 8 1 8 3 0 5 1 8 8]

;
  )

File: python/data_structures/union_find/quick_union.py

class QuickUnion:
    def __init__(self, N: int) -> int:
        self.ids = [i for i in range(N)]

    def root(self, p: int) -> int:
        while p != self.ids[p]:
            p = self.ids[p]
        return p

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def union(self, p, q):
        proot = self.root(p)
        qroot = self.root(q)
        self.ids[proot] = qroot

if __name__ == '__main__':
    qu = QuickUnion(10)
    qu.union(4, 3)
    qu.union(3, 8)
    qu.union(6, 5)
    qu.union(9, 4)
    qu.union(2, 1)
    qu.union(5, 0)
    qu.union(7, 2)
    qu.union(6, 1)
    qu.union(7, 3)
    print(qu.ids)
  #=> [1 8 1 8 3 0 5 1 8 8]
